resources/2024-07-secretstore-feature-spec.md

```diff
@@ -0,0 +1,229 @@
+
+As an application developer, I cannot reference a `secretStores` resource in the `Applications.Core/containers`, `Applications.Core/extenders`, `Applications.Core/volumes`, `Applications.Datastores/*`, and `Applications.Messaging/*` resource types to securely manage secrets for use in my application. Instead, I'm having to store my secrets in plain text within the `properties.secrets` field for each resource. This is a critical gap in the secret management capabilities of Radius that is hindering my ability to securely manage secrets for use in my containers.
+
+## Desired customer experience outcome
+<!-- <Write this as an “I statement” that expresses the new capability from customer perspective … i.e. After this scenario is implemented “I can do, x, y, z, steps in cloud native app developer and seamlessly blah blah blah …. As a result <summarize positive impact on your work / business>  -->
+
+As an operator, I can define an `Applications.Core/secretStores` resource and deploy it along with an Environment so that the developers I support can securely leverage secrets I manage on their behalf for use in their application resources.
+
+As a developer, I can reference an `Applications.Core/secretStores` in my `Applications.Core/containers` resource definition so that Radius will inject secrets as environment variables into my container at deploy time so that credentials can be provided to the container for authentication, etc.
+
+As a developer, I can reference an `Applications.Core/secretStores` in my `Applications.Extenders` or `Applications.Volumes` resource definition so that I no longer have to store secrets as plain text in the `properties.secrets` field of my resource for authentication, etc.
+
+As a developer, I can reference an `Applications.Core/secretStores` in my `Applications.Datastores/*` or `Applications.Messaging/*` resource definition so that I no longer have to store secrets as plain text in the `properties.secrets` field of my resource for authentication, etc.
+
+### Detailed Customer Experience
+ <!-- <List of steps the customer goes through from the start to the end of the scenario to provide more detailed view of exactly what the user is able to do given the new capabilities>  -->
+<!-- Step 1
+Step 2
+… -->
+
+Step 1: Operator creates and deploys `Applications.Core/secretStores` resources containing the secret data required for authenticating into resources that their developers may use.
+
+```bicep
+resource authcreds 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'authcreds'
+  properties:{
+    application: application
+    type: 'credentials' // this can change based on detailed tech design
```

Comment:
Today the implementation is actually an ENUM with possible values `["generic", "certificate"]`

There was some related discussion in https://github.com/radius-project/design-notes/pull/53 where the decision was to extend this ENUM with additional values for their use cases. We have an opportunity to propose a change here to make this `type` field free form to enable better extensibility going forward so that we don't have to keep updating the API each time. What are your thoughts here @superbeeny?

Reply:
We picked `generic` over `opaque` because we found it much clearer.

Reply:
Something else that's needed here is define a naming scheme for the values of `type`. What should the names be for AzWI? for AWS IRSA? For the secret created by an `Applications.Datastores/redisCaches` resource?

Reply:
this has been defined here already: https://github.com/radius-project/radius/pull/7816